### ISSUE TITLE
error referencing function defined using lazy specifier

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon-compiler#2077]
I'm not sure it ever makes sense to have a variable parameter on a function defined using the lazy specifier, but doing so can result in a backend error:

```ceylon
    Boolean foo<Element>(variable Element x) => x exists;
    
    print (foo<Anything>(null)); // ok
    print (foo<Anything>("")); // ok
    value fooAnything = foo<Anything>; // Cannot find symbol: class Element
```

**Comment by quintesse** [comment]
Indeed, doesn't really seem to make much sense. But on the other had I also don't see a reason to forbid it in the spec for example, so we should probably just fix it.
Not really high prio so assigning to 1.3

**Comment by lucaswerkmeister** [comment]
Well, it’s definitely possible to write legal code that actually varies the parameter:
```ceylon
Boolean complicatedFalse<Element>(variable Element? x) => ((x = null) of Element?) exists;
```
With let expressions you could also get the old value, then assign to it, then use both the old and the new value. Probably still doesn’t make sense, but I don’t think it should be illegal either.

**Comment by jvasileff** [comment]
Oh, nice @lucaswerkmeister! I forgot about that. Now I can *finally* write a memoizer without ugly braces!

```ceylon
Result memoize<Result>
        (Result() f, variable Result? memo=null)()
        given Result satisfies Object
    =>  memo = memo else f();
```

**Comment by lucaswerkmeister** [comment]
Wow, that’s fancy.

**Comment by quintesse** [comment]
I might be completely wrong but I doubt that will work, the `memo` wouldn't be stored between calls to the function AFAIK.

**Comment by lucaswerkmeister** [comment]
@quintesse seems to work as intended. [gist](https://gist.github.com/lucaswerkmeister/84fb2b308421711e2e95)

**Comment by quintesse** [comment]
Color me surprised :)

**Comment by gavinking** [comment]
Haha, cool :)

**Comment by quintesse** [comment]
I hadn't really noticed the extra pair of parenthesis, nice :)
